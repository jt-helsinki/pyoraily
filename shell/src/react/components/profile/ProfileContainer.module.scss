.fullWidthButton {
  height: var(--s-xl);
  width: 100% !important;
  border-radius: 0 !important;
  max-width: unset !important;
  max-height: unset !important;
}

div.userProfileBody {
  min-height: 100%;
  height: auto !important;
  height: 100%;
  margin-bottom: calc(var(--s-xl) * -1);
}

div.userProfileAddressLine {
  line-height: var(--line-height-root);
}

div.profileContainer {
  display: grid;
  column-gap: var(--s-b);
  row-gap: var(--s-b);

  div.profileLine {
    grid-column-start: 1;
    grid-column-end: 8;
  }

  div.profile-column-left {
    grid-column-start: 1;
    grid-column-end: 1;
  }

  div.profileColumnRight {
    grid-column-start: 2;
    grid-column-end: 8;
  }
}
