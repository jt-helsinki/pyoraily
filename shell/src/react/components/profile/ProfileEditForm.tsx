/*
 *
 * MIT License.
 *
 */
import * as UserUtils from 'pyoraily-shared-frontend/utils/UserUtils';
import { MIN_YEAR, ProfileEditFormModelSchema } from '@src/model/schema/ProfileFormSchemas';
import { User } from 'pyoraily-shared-frontend/model/User';
import { UserRole } from 'pyoraily-shared-frontend/model/UserRole';
import { Formik, FormikProps } from 'formik';
import * as React from 'react';
import { Button, Dimmer, Dropdown, Form, Header, Label, Loader, Modal, Radio } from 'semantic-ui-react';

interface Props {
  user: User;

  userRoles: UserRole[];

  isSubmitting: boolean;

  onSave: (user: User) => void;

  onCancel: () => void;
}

const years = Array.from({ length: 65 }, (e, i) => i + MIN_YEAR)
  .map((year) => ({
    key: year,
    text: year,
    value: year,
  }))
  .reverse();

export const ProfileEditForm = (props: Props): React.ReactElement => {
  const isAthlete: boolean = UserUtils.doUserRolesIntersectRequiredUserRoles(props.userRoles, [UserRole.ATHLETE]);
  return (
    <>
      <Dimmer active={props.isSubmitting}>
        <Loader>Saving</Loader>
      </Dimmer>
      <Header as="h3">Update your profile information.</Header>
      <Formik
        validateOnChange={true}
        validateOnBlur={true}
        enableReinitialize={true}
        initialValues={props.user}
        validationSchema={ProfileEditFormModelSchema}
        onSubmit={(values: User): void => {
          props.onSave(values);
        }}>
        {(formik: FormikProps<User>): React.ReactElement => (
          <Form id="profile-update-form">
            <Form.Field>
              <label>First Name</label>
              <Form.Input
                type="text"
                required={true}
                id="profile-update-form-first-name"
                name="firstName"
                value={formik.values.firstName}
                disabled={props.isSubmitting}
                onChange={formik.handleChange}
                error={
                  formik.errors.firstName && (formik.submitCount > 0 || formik.touched.firstName)
                    ? { content: formik.errors.firstName, pointing: 'above' }
                    : false
                }
              />
            </Form.Field>
            <Form.Field>
              <label>Last Name</label>
              <Form.Input
                type="text"
                required={true}
                id="profile-update-form-last-name"
                name="lastName"
                value={formik.values.lastName}
                disabled={props.isSubmitting}
                onChange={formik.handleChange}
                error={
                  formik.errors.lastName && (formik.submitCount > 0 || formik.touched.lastName)
                    ? { content: formik.errors.lastName, pointing: 'above' }
                    : false
                }
              />
            </Form.Field>
            {isAthlete ? (
              <>
                <Form.Group inline={true} unstackable={false}>
                  <label>Gender</label>
                  <Form.Field
                    control={Radio}
                    label="Male"
                    value="male"
                    name="gender"
                    disabled={props.isSubmitting}
                    checked={formik.values.gender === 'male'}
                    onClick={() => formik.setFieldValue('gender', 'male', false)}
                    error={!!(formik.errors.gender && (formik.submitCount > 0 || formik.touched.gender))}
                  />
                  <Form.Field
                    control={Radio}
                    label="Female"
                    value="female"
                    name="gender"
                    disabled={props.isSubmitting}
                    checked={formik.values.gender === 'female'}
                    onClick={() => formik.setFieldValue('gender', 'female', false)}
                    error={!!(formik.errors.gender && (formik.submitCount > 0 || formik.touched.gender))}
                  />
                  {formik.errors.gender && (formik.submitCount > 0 || formik.touched.gender) ? (
                    <Label prompt pointing="above">
                      {formik.errors.gender}
                    </Label>
                  ) : null}
                </Form.Group>
                <Form.Field>
                  <label>Year of Birth</label>
                  <Dropdown
                    placeholder="Year of Birth"
                    selection
                    options={years}
                    defaultValue={formik.values.yearOfBirth}
                    disabled={props.isSubmitting}
                    onChange={(event: any, data: any) => formik.setFieldValue('yearOfBirth', data.value, false)}
                    error={!!(formik.errors.yearOfBirth && (formik.submitCount > 0 || formik.touched.yearOfBirth))}
                  />
                  {formik.errors.yearOfBirth && (formik.submitCount > 0 || formik.touched.yearOfBirth) ? (
                    <Label prompt pointing="above">
                      {formik.errors.yearOfBirth}
                    </Label>
                  ) : null}
                </Form.Field>
              </>
            ) : null}
            <Form.Field>
              <label>Email</label>
              <Form.Input
                type="text"
                required={true}
                id="profile-update-form-email"
                name="email"
                value={formik.values.email}
                disabled={props.isSubmitting}
                onChange={formik.handleChange}
                error={
                  formik.errors.email && (formik.submitCount > 0 || formik.touched.email)
                    ? { content: formik.errors.email, pointing: 'above' }
                    : false
                }
              />
            </Form.Field>
            <Form.Field>
              <label>UCI ID</label>
              <Form.Input
                type="number"
                required={true}
                id="profile-update-form-uci-id"
                name="uciId"
                value={formik.values.uciId}
                disabled={props.isSubmitting}
                onChange={formik.handleChange}
                error={
                  formik.errors.uciId && (formik.submitCount > 0 || formik.touched.uciId)
                    ? { content: formik.errors.uciId, pointing: 'above' }
                    : false
                }
              />
            </Form.Field>
            <Button secondary={true} onClick={props.onCancel}>
              Cancel
            </Button>
            <Button primary={true} onClick={formik.submitForm}>
              Save
            </Button>
          </Form>
        )}
      </Formik>
    </>
  );
};
