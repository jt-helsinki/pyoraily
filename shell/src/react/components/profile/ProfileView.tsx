/*
 *
 * MIT License.
 *
 */
import * as UserUtils from 'pyoraily-shared-frontend/utils/UserUtils';
import { User } from 'pyoraily-shared-frontend/model/User';
import { UserRole } from 'pyoraily-shared-frontend/model/UserRole';
import * as UserComponentUtils from '@src/react/components/common/UserComponentUtils';
import * as React from 'react';
import { Button, Grid, Header, Icon } from 'semantic-ui-react';
import { stringToSentenceCase } from 'pyoraily-shared-frontend/utils/StringUtils';

interface Props {
  user: User;

  userRoles: UserRole[];

  onEdit: () => void;
}

export const ProfileView = (props: Props): React.ReactElement => {
  const isAthlete: boolean = UserUtils.doUserRolesIntersectRequiredUserRoles(props.userRoles, [UserRole.ATHLETE]);
  return (
    <>
      <Header as="h3">Your profile information.</Header>
      <Grid>
        <Grid.Row>
          <Grid.Column>{UserComponentUtils.userRolesToTags(props.userRoles)}</Grid.Column>
        </Grid.Row>
        <Grid.Row>
          <Grid.Column>
            <p>
              <Icon name="user outline" />
            </p>
          </Grid.Column>
          <Grid.Column width={4}>
            <p>{props.user.firstName || 'No first name set.'}</p>
          </Grid.Column>
        </Grid.Row>
        <Grid.Row>
          <Grid.Column>
            <p>
              <Icon name="user" />
            </p>
          </Grid.Column>
          <Grid.Column width={4}>
            <p>{props.user.lastName || 'No last name set.'}</p>
          </Grid.Column>
        </Grid.Row>
        {isAthlete ? (
          <>
            <Grid.Row>
              <Grid.Column>
                <p>
                  <Icon name="heterosexual" />
                </p>
              </Grid.Column>
              <Grid.Column width={4}>
                <p>{stringToSentenceCase((props.user.gender && props.user.gender) || 'No gender set.')}</p>
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column>
                <p>
                  <Icon name="birthday cake" />
                </p>
              </Grid.Column>
              <Grid.Column width={4}>
                <p>{props.user.yearOfBirth || 'No year of birth set.'}</p>
              </Grid.Column>
            </Grid.Row>
          </>
        ) : null}
        <Grid.Row>
          <Grid.Column>
            <p>
              <Icon name="mail" />
            </p>
          </Grid.Column>
          <Grid.Column width={4}>
            <p>{props.user.email || 'No email set.'}</p>
          </Grid.Column>
        </Grid.Row>
        <Grid.Row>
          <Grid.Column>
            <p>
              <Icon name="id badge" />
            </p>
          </Grid.Column>
          <Grid.Column width={4}>
            <p>{props.user.uciId || 'No UCI ID set.'}</p>
          </Grid.Column>
        </Grid.Row>
        <Grid.Row>
          <Grid.Column>
            <Button primary={true} onClick={props.onEdit}>
              Update
            </Button>
          </Grid.Column>
        </Grid.Row>
      </Grid>
    </>
  );
};
