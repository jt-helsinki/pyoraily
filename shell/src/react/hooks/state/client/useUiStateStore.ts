/*
 *
 * MIT License.
 *
 */
import { UiState } from '@src/state/client/state/UiState';
import { UiStateManager } from '@src/state/client/UiStateManager';
import { StateSelector } from 'pyoraily-shared-frontend/state/lib/StateSelector';

/**
 * @see useUiStateStore()
 * @param {(state: UiState) => any} selector a custom selector to return specific slices of the state.
 * @return {UiState} The underlying instance of the UiState
 */
export const useUiStateStoreWithSelector = (selector: (state: UiState) => any): UiState =>
  UiStateManager.store()(selector) as any;

/**
 * A React hook to return an implementation of the underlying UiState store as defined by the UiState.
 *
 * @param {keyof UiState} propertiesToPicks the properties of the state to return to the caller.
 *
 *   1. If just one property is specified then only the raw value of the property will be selected.
 *   2. If multiple properties are specified then each property/value will be selected. Note: This condition will always
 *   cause a state update regardless of whether the state has updated or njot.
 *   3. If no properties are specified, the whole state object will be selected.
 * @return {UiState} The underlying instance of the UiState
 */
export const useUiStateStore = (...propertiesToPick: (keyof UiState)[]): UiState | any => {
  const selector: (state: UiState) => UiState = StateSelector.selector<UiState>(propertiesToPick);
  return useUiStateStoreWithSelector(selector) as any;
};
