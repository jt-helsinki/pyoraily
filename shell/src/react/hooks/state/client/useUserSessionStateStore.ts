/*
 *
 * MIT License.
 *
 */
import * as UserUtils from 'pyoraily-shared-frontend/utils/UserUtils';
import { UserRole } from 'pyoraily-shared-frontend/model/UserRole';
import { UserSessionState } from '@src/state/client/state/UserSessionState';
import { UserSessionStateManager } from '@src/state/client/UserSessionStateManager';
import { StateSelector } from 'pyoraily-shared-frontend/state/lib/StateSelector';

/**
 * @see useUserSessionStateStore()
 * @param {(state: UserSessionState) => any} selector a custom selector to return specific slices of the state.
 * @return {UserSessionState} The underlying instance of the UserSessionState
 */
export const useUserSessionStateStoreWithSelector = (
  selector: (state: UserSessionState) => any
): UserSessionState | any => UserSessionStateManager.store()(selector) as any;

/**
 * A React hook to return an implementation of the underlying UserSessionState store as defined by the UserSessionState.
 *
 * @param {keyof UserSessionState} propertiesToPicks the properties of the state to return to the caller.
 *
 *   1. If just one property is specified then only the raw value of the property will be selected.
 *   2. If multiple properties are specified then each property/value will be selected. Note: This condition will always
 *   cause a state update regardless of whether the state has updated or njot.
 *   3. If no properties are specified, the whole state object will be selected.
 * @return {UserSessionState} The underlying instance of the UserSessionState
 */
export const useUserSessionStateStore = (...propertiesToPick: (keyof UserSessionState)[]): UserSessionState | any => {
  const selector: (state: UserSessionState) => UserSessionState =
    StateSelector.selector<UserSessionState>(propertiesToPick);
  return useUserSessionStateStoreWithSelector(selector) as any;
};

/**
 * Returns the user roles contained in the UserSessionState
 * @return {string} The current user's user roles as a UserRole[].
 */
export const useUserRolesFromUseUserSessionStateStore = (): UserRole[] =>
  useUserSessionStateStoreWithSelector((state: UserSessionState): UserRole[] => state.userRoles);

/**
 * Based on the userRoles in the current user's session, does the user have permission tagged with a particular ID.
 * @return {boolean} Whether (or not) the current usercan access a resource.
 */
export const useCanUserAccessResource = (userRoles: UserRole[] | UserRole): boolean => {
  const userRolesForUser: UserRole[] = useUserRolesFromUseUserSessionStateStore();
  return UserUtils.doUserRolesIntersectRequiredUserRoles(userRolesForUser, userRoles);
};
