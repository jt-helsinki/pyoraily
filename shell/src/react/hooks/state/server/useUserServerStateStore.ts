/*
 *
 * MIT License.
 *
 */
import { User } from 'pyoraily-shared-frontend/model/User';
import { UserRole } from 'pyoraily-shared-frontend/model/UserRole';
import { useUserSessionStateStore } from '@src/react/hooks/state/client/useUserSessionStateStore';
import * as UserServerStateManager from '@src/state/server/UserServerStateManager';
import { useMutation, useQuery } from '@tanstack/react-query';
import {
  UseMutationOptions,
  UseMutationResult,
  UseQueryOptions,
  UseQueryResult,
} from '@tanstack/react-query/src/types';

export const useFetchUsersServerStateStore = (
  reactQueryOptions: UseQueryOptions = {}
): UseQueryResult<User[], Error> => {
  const enabled: boolean = useUserSessionStateStore('authenticated');
  return useQuery(
    UserServerStateManager.fetchUsers({
      enabled,
      ...reactQueryOptions,
    }) as any
  );
};

export const useFetchUserRolesServerStateStore = (
  reactQueryOptions: UseQueryOptions = {}
): UseQueryResult<UserRole[], Error> => {
  const enabled: boolean = useUserSessionStateStore('authenticated');
  return useQuery(UserServerStateManager.fetchUserRoles({ enabled, ...reactQueryOptions }) as any);
};

export const useFetchProfileServerStateStore = (
  reactQueryOptions: UseQueryOptions = {}
): UseQueryResult<User, Error> => {
  const enabled: boolean = useUserSessionStateStore('authenticated');
  return useQuery(UserServerStateManager.fetchProfile({ enabled, ...reactQueryOptions }) as any);
};

export const useUpdateProfileServerStateStore = (
  reactQueryOptions: UseMutationOptions = {}
): UseMutationResult<User, Error, Partial<User>, any> =>
  useMutation(UserServerStateManager.updateProfile({ ...reactQueryOptions }) as any);

export const useUpdateUserServerStateStore = (
  reactQueryOptions: UseMutationOptions = {}
): UseMutationResult<User, Error, Partial<User>, any> =>
  useMutation(UserServerStateManager.updateUser({ ...reactQueryOptions }) as any);

export const useCreateUserServerStateStore = (
  reactQueryOptions: UseMutationOptions = {}
): UseMutationResult<User, Error, User, any> =>
  useMutation(UserServerStateManager.createUser({ ...reactQueryOptions }) as any);
