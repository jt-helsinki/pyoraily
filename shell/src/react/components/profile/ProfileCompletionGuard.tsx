/*
 *
 * MIT License.
 *
 */

import { useFetchProfileServerStateStore } from '@src/react/hooks/state/server/useUserServerStateStore';
import { PROTECTED_ROUTE_PATH_NODES } from '@src/react/routes/RouteConfig';
import * as RouteUtils from '@src/react/routes/RouteUtils';
import * as React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { Button, Dimmer, Loader, Message } from 'semantic-ui-react';

const profilePath: string = RouteUtils.removeWildcardFromPath(PROTECTED_ROUTE_PATH_NODES.profile.path);

export const ProfileCompletionGuard: React.FunctionComponent<React.PropsWithChildren> = (
  props: React.PropsWithChildren
): React.ReactElement => {
  const { data: user, isFetching } = useFetchProfileServerStateStore();
  const navigate = useNavigate();
  const location = useLocation();

  if (isFetching) {
    return (
      <Dimmer active={isFetching}>
        <Loader>Fetching your profile</Loader>
      </Dimmer>
    );
  }

  return (user?.firstName && user?.lastName) || location.pathname === profilePath ? (
    <>{props.children}</>
  ) : (
    <Message warning={true}>
      <Message.Header>Profile not setup.</Message.Header>
      <p>Please update your profile before using this application.</p>
      <Button onClick={() => navigate(profilePath)}>Profile</Button>
    </Message>
  );
};
