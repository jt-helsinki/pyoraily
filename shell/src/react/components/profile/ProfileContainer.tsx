/*
 *
 * MIT License.
 *
 */
/*
 *
 * MIT License.
 *
 */
import { User } from 'pyoraily-shared-frontend/model/User';
import { UserRole } from 'pyoraily-shared-frontend/model/UserRole';
import { ProfileEditForm } from '@src/react/components/profile/ProfileEditForm';
import { ProfileView } from '@src/react/components/profile/ProfileView';
import { useUserRolesFromUseUserSessionStateStore } from '@src/react/hooks/state/client/useUserSessionStateStore';
import {
  useFetchProfileServerStateStore,
  useUpdateProfileServerStateStore,
} from '@src/react/hooks/state/server/useUserServerStateStore';
import { NotificationKey } from '@src/react/types/NotificationKey';
import * as UiService from '@src/services/UiService';
import * as React from 'react';
import { Dimmer, Loader, Modal } from 'semantic-ui-react';

enum ComponentMode {
  View,
  EditProfile,
}

export const ProfileContainer: React.FunctionComponent = (): React.ReactElement => {
  const [error, setError] = React.useState<string>('');
  const [isSubmitting, setIsSubmitting] = React.useState<boolean>(false);
  const [componentMode, setComponentMode] = React.useState<ComponentMode>(ComponentMode.View);
  const userRoles: UserRole[] = useUserRolesFromUseUserSessionStateStore();
  const { data: user, refetch, isFetching } = useFetchProfileServerStateStore();
  const { mutateAsync: updateProfileMutate } = useUpdateProfileServerStateStore();

  const onEditCancel = (): void => {
    setComponentMode(ComponentMode.View);
    setError(null as any);
    setIsSubmitting(false);
  };

  const onEdit = (): void => {
    setComponentMode(ComponentMode.EditProfile);
  };

  const onEditSave = (profile: User): void => {
    setError(null as any);
    setIsSubmitting(true);
    updateProfileMutate(profile)
      .then((savedUser: User) => {
        UiService.notificationToasterSuccess(
          NotificationKey.USERS_EMAIL_UPDATED,
          `User ${savedUser.firstName} ${savedUser.lastName} <${savedUser.email}> saved`
        );
        refetch().then(() => onEditCancel());
      })
      .catch((responseError: string) => {
        setError(responseError);
        setIsSubmitting(false);
      });
  };

  if (isFetching) {
    return (
      <Dimmer active={isFetching}>
        <Loader>Fetching profile</Loader>
      </Dimmer>
    );
  }

  return (
    <>
      {componentMode === ComponentMode.View ? (
        <ProfileView user={user as User} userRoles={userRoles} onEdit={onEdit} />
      ) : (
        <ProfileEditForm
          user={user as User}
          userRoles={userRoles}
          isSubmitting={isSubmitting}
          onSave={onEditSave}
          onCancel={onEditCancel}
        />
      )}
    </>
  );
};
