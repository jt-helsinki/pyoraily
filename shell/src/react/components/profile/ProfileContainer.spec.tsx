import { ProfileContainer } from '@src/react/components/profile/ProfileContainer';
import * as UserServerStateStore from '@src/react/hooks/state/server/useUserServerStateStore';
import { UiStateManager } from '@src/state/client/UiStateManager';
import { UserSessionStateManager } from '@src/state/client/UserSessionStateManager';
import { mockFetchProfileResponseDate } from '@src/tests/mocks/mockUserServiceResponses';
import { renderSUI, userSessionSetup } from '@src/tests/utils';
import { screen } from '@testing-library/react';
import * as React from 'react';
import * as ReactRouter from 'react-router';

describe('Profile', () => {
  beforeEach(() => {
    userSessionSetup();

    const mockedUsedNavigate = jest.fn();
    jest.spyOn(ReactRouter, 'useNavigate').mockImplementation(mockedUsedNavigate);
    jest
      .spyOn(UserServerStateStore, 'useFetchProfileServerStateStore')
      .mockImplementation(() => ({ data: mockFetchProfileResponseDate } as any));
  });

  afterAll(() => {
    UserSessionStateManager.resetToNew();
    UiStateManager.resetToNew();
    jest.resetAllMocks();
  });

  it('should match the snapshot', () => {
    const { container } = renderSUI(<ProfileContainer />);
    expect(container).toMatchSnapshot();
  });

  it('should display user email and address', async () => {
    renderSUI(<ProfileContainer />);
    expect(screen.getByText('Skid')).toBeInTheDocument();
    expect(screen.getByText('Mark')).toBeInTheDocument();
    expect(screen.getByText('male')).toBeInTheDocument();
    expect(screen.getByText('2000')).toBeInTheDocument();
    expect(screen.getByText('No UCI ID set.')).toBeInTheDocument();
    expect(screen.getByText('test@test.com')).toBeInTheDocument();
  });
});
