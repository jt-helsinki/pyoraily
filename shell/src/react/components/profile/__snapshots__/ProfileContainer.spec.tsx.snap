// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile should match the snapshot 1`] = `
<div>
  <div
    class="ui grid"
  >
    <div
      class="row"
    >
      <div
        class="column"
      >
        <div
          class="ui grey tiny label"
        >
          HPY
        </div>
        <div
          class="ui grey tiny label"
        >
          SP
        </div>
        <div
          class="ui grey tiny label"
        >
          ATHLETE
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="column"
      >
        <p>
          <i
            aria-hidden="true"
            class="user outline icon"
          />
        </p>
      </div>
      <div
        class="four wide column"
      >
        <p>
          Skid
        </p>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="column"
      >
        <p>
          <i
            aria-hidden="true"
            class="user icon"
          />
        </p>
      </div>
      <div
        class="four wide column"
      >
        <p>
          Mark
        </p>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="column"
      >
        <p>
          <i
            aria-hidden="true"
            class="heterosexual icon"
          />
        </p>
      </div>
      <div
        class="four wide column"
      >
        <p>
          male
        </p>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="column"
      >
        <p>
          <i
            aria-hidden="true"
            class="birthday cake icon"
          />
        </p>
      </div>
      <div
        class="four wide column"
      >
        <p>
          2000
        </p>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="column"
      >
        <p>
          <i
            aria-hidden="true"
            class="mail icon"
          />
        </p>
      </div>
      <div
        class="four wide column"
      >
        <p>
          test@test.com
        </p>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="column"
      >
        <p>
          <i
            aria-hidden="true"
            class="id badge icon"
          />
        </p>
      </div>
      <div
        class="four wide column"
      >
        <p>
          No UCI ID set.
        </p>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="column"
      >
        <button
          class="ui primary button"
        >
          Update
        </button>
      </div>
    </div>
  </div>
</div>
`;
